server/src/pages/shot/editor.js

```diff
@@ -213,319 +144,64 @@ exports.Editor = class Editor extends React.Component {
-      </Localized>
-      <Localized id="annotationCropCancelButton">
-        <button className={`button transparent cancel-crop`} id="cancel-crop" onClick={this.onClickCancelCrop.bind(this)} title="Cancel selection">Cancel</button>
-      </Localized>
-    </div></div>;
+    this.setState({tool: "cropTool"});
     sendEvent("crop-select", "annotation-toolbar");
   }
 
-  onClickConfirmCrop() {
-    const x1 = Math.max(selectedPos.left, 0);
-    const x2 = Math.min(selectedPos.right, this.canvasWidth);
-    const y1 = Math.max(selectedPos.top, 0);
-    const y2 = Math.min(selectedPos.bottom, this.canvasHeight);
-    const cropWidth = Math.floor(x2 - x1);
-    const cropHeight = Math.floor(y2 - y1);
-    if (!selectedPos.width || !selectedPos.height || (this.canvasHeight === cropHeight) && (this.canvasWidth === cropWidth)) {
-      this.removeCropBox();
-      this.cropToolBar = null;
+  onDrawingUpdate(affectedArea, incomingCanvas) {
+    if (!affectedArea || !incomingCanvas) {
+      return;
+    }
+
+    this.imageContext.drawImage(incomingCanvas,
+      affectedArea.left * this.devicePixelRatio,
+      affectedArea.top * this.devicePixelRatio,
+      affectedArea.width * this.devicePixelRatio,
+      affectedArea.height * this.devicePixelRatio,
+      affectedArea.left, affectedArea.top, affectedArea.width, affectedArea.height);
+  }
+
+  onCropUpdate(affectedArea, incomingCanvas) {
+    if (!affectedArea || !incomingCanvas) {
       this.setState({tool: this.previousTool});
       return;
     }
-    const croppedImage = document.createElement("canvas");
-    croppedImage.width = cropWidth * this.devicePixelRatio;
-    croppedImage.height = cropHeight * this.devicePixelRatio;
-    const croppedContext = croppedImage.getContext("2d");
-    croppedContext.drawImage(this.imageCanvas, x1 * this.devicePixelRatio, y1 * this.devicePixelRatio, croppedImage.width, croppedImage.height, 0, 0, croppedImage.width, croppedImage.height);
+
     const img = new Image();
     img.crossOrigin = "Anonymous";
-    const width = cropWidth;
-    const height = cropHeight;
     img.onload = () => {
       imageContext.scale(this.devicePixelRatio, this.devicePixelRatio);
-      imageContext.drawImage(img, 0, 0, width, height);
+      imageContext.drawImage(img, 0, 0, affectedArea.width, affectedArea.height);
     };
-    this.canvasWidth = cropWidth;
-    this.canvasHeight = cropHeight;
+    this.canvasWidth = affectedArea.width;
+    this.canvasHeight = affectedArea.height;
     const imageContext = this.imageCanvas.getContext("2d");
     this.imageContext = imageContext;
-    img.src = croppedImage.toDataURL("image/png");
-    this.removeCropBox();
-    this.cropToolBar = null;
+    img.src = incomingCanvas.toDataURL("image/png");
     this.setState({tool: this.previousTool});
-    imageIsCropped = true;
-    sendEvent("confirm-crop", "crop-toolbar");
   }
```

Comment:
This replaces the image in the base canvas with the one in the incoming canvas.  For undo/redo, we insert a copy of the whole image as a key frame in the history.